libs/vkd3d/va_map.c

```diff
@@ -0,0 +1,381 @@
+            }
+            else
+            {
+                vkd3d_atomic_uint64_store_explicit(&block->l.va, max_va, vkd3d_memory_order_relaxed);
+                vkd3d_atomic_ptr_store_explicit(&block->l.resource, resource, vkd3d_memory_order_relaxed);
+            }
+
+            block_va += VKD3D_VA_BLOCK_SIZE;
+        }
+    }
+    else
+    {
+        pthread_mutex_lock(&va_map->mutex);
+
+        if (!vkd3d_va_map_find_small_entry(va_map, resource->va, &index))
+        {
+            vkd3d_array_reserve((void**)&va_map->small_entries, &va_map->small_entries_size,
+                    va_map->small_entries_count + 1, sizeof(*va_map->small_entries));
+
+            memmove(&va_map->small_entries[index + 1], &va_map->small_entries[index],
+                    sizeof(*va_map->small_entries) * (va_map->small_entries_count - index));
+
+            va_map->small_entries[index] = resource;
+            va_map->small_entries_count += 1;
+        }
+
+        pthread_mutex_unlock(&va_map->mutex);
+    }
+}
+
+void vkd3d_va_map_remove(struct vkd3d_va_map *va_map, const struct vkd3d_unique_resource *resource)
+{
+    VkDeviceAddress block_va, min_va, max_va;
+    struct vkd3d_va_block *block;
+    size_t index;
+
+    if (resource->size >= VKD3D_VA_BLOCK_SIZE)
+    {
+        min_va = resource->va;
+        max_va = resource->va + resource->size;
+        block_va = min_va & ~VKD3D_VA_LO_MASK;
+
+        while (block_va < max_va)
+        {
+            block = vkd3d_va_map_get_block(va_map, block_va);
+
+            if (vkd3d_atomic_ptr_load_explicit(&block->l.resource, vkd3d_memory_order_relaxed) == resource)
+            {
+                vkd3d_atomic_uint64_store_explicit(&block->l.va, 0, vkd3d_memory_order_relaxed);
+                vkd3d_atomic_ptr_store_explicit(&block->l.resource, NULL, vkd3d_memory_order_relaxed);
+            }
+            else if (vkd3d_atomic_ptr_load_explicit(&block->r.resource, vkd3d_memory_order_relaxed) == resource)
+            {
+                vkd3d_atomic_uint64_store_explicit(&block->r.va, 0, vkd3d_memory_order_relaxed);
+                vkd3d_atomic_ptr_store_explicit(&block->r.resource, NULL, vkd3d_memory_order_relaxed);
+            }
+
+            block_va += VKD3D_VA_BLOCK_SIZE;
+        }
+    }
+    else
+    {
+        pthread_mutex_lock(&va_map->mutex);
+
+        if (vkd3d_va_map_find_small_entry(va_map, resource->va, &index) == resource)
+        {
+            va_map->small_entries_count -= 1;
+
+            memmove(&va_map->small_entries[index], &va_map->small_entries[index + 1],
+                    sizeof(*va_map->small_entries) * (va_map->small_entries_count - index));
+        }
+
+        pthread_mutex_unlock(&va_map->mutex);
+    }
+}
+
+const struct vkd3d_unique_resource *vkd3d_va_map_deref(struct vkd3d_va_map *va_map, VkDeviceAddress va)
+{
+    const struct vkd3d_va_block *block = vkd3d_va_map_find_block(va_map, va);
+    const struct vkd3d_unique_resource *resource = NULL;
```

Comment:
Looks like declare after statement to me.

Reply:
Is it though? GCC isn't complaining and we do this sort of thing all the time (impl_from_ID3D12xxxx etc).

Reply:
Ok, no complaints then.